### ISSUE TITLE
Relocatable store support

### ISSUE THREAD

**Issue by gsakkis** [zarr-developers/zarr-python#548]
Although zarr supports pickling, currently unpickling fails if the underlying store changes, e.g when moving the directory of a `DirectoryStore` or the file of a `ZipStore`. As mentioned in the docs:
>  The only thing that is pickled is the necessary parameters to allow the store to re-open any underlying files or databases upon being unpickled.
How can I add support for unpickling from a different store than the one used for pickling?

**Comment by alimanfoo** [comment]
Hi @gsakkis, would you be able to say a bit more about the use case for this? Would really help to get a sense of the problem you're trying to solve.

**Comment by gsakkis** [comment]
Hi @alimanfoo, the use case is the following:
- On server A I generate some zarr arrays, save them in a zip file and pickle the `zarr.Array` instance(s):
```py
In [1]: import pickle, zarr, numpy as np                                                                                                                         

In [2]: z = zarr.open('data/walnuts.zarr.zip', mode='w', shape=100000, dtype='i8')                                                                               

In [3]: z[:] = np.arange(100000)                                                                                                                                 

In [4]: with open('data/walnuts.pickle', 'wb') as f: 
   ...:     pickle.dump(z, f) 
   ...:                                                                                                                                                          
```
- Then I send these two files in server B (or just move them to a different directory in the same server) and try to unpickle the pickled array:
```py
In [5]: with open('/tmp/walnuts.pickle', 'rb') as f: 
   ...:     pickle.load(f) 
KeyError                                  Traceback (most recent call last)
...
KeyError: "There is no item named '.zarray' in the archive"

During handling of the above exception, another exception occurred:

ValueError                                Traceback (most recent call last)
...
ValueError: array not found at path ''
```
Hope it's clearer now.

By the way I have come up with some solution, I'll open a PR soon.

**Comment by alimanfoo** [comment]
Thanks a lot @gsakkis, much appreciated.

Sorry to be cautious here, but could you do this without pickling? I see in #549 you've made it possible to set a different store on the array, but if you're going to set a different store, couldn't you just instantiate a new array instead?

E.g., in server A, run something like:

```
z = zarr.open('data/walnuts.zarr.zip', mode='w', shape=100000, dtype='i8') 
z[:] = np.arange(100000)
```

If you are copying the data file to a location on server B which has a different path from server A, e.g., "some/other/path/walnuts.zarr.zip", then could you just run this on server B:

```
z = zarr.open('some/other/path/walnuts.zarr.zip', mode='r')
```

What do you gain by pickling, sending, then unpickling the array between the two servers?

Again sorry to be cautious, but this a critical area of the code and want to be sure it's worth it before considering changes.

**Comment by gsakkis** [comment]
@alimanfoo no problem, I understand the cautiousness. The example above was minimal to describe the issue, and indeed for a single array there's no need for a pickle. The actual use case is a deeply nested structure (think of dict of dicts/lists of ...) where some values are currently numpy arrays. This structure is what's pickled and sent back. So what I want to do is replace the numpy arrays with zarr and send back two files, the pickled nested structure and a zip file of the actual zarrified arrays.

**Comment by jakirkham** [comment]
Are you familiar with the `.update(...)` method on `MutableMapping`s? This would allow you to copy the contents of one store into another.

**Comment by gsakkis** [comment]
@jakirkham not sure how this is related; even if `update` worked recursively on nested mappings and lists (it doesn't), on the remote side I can't even open the original pickled object to copy its contents to a different store.

**Comment by jakirkham** [comment]
It would allow you to copy the underlying data from one store to another. Say from a `DirectoryStore` to a `ZipStore`.

**Comment by gsakkis** [comment]
I can't access the original `DirectoryStore`; the copying takes place on a different server.

**Comment by jakirkham** [comment]
I’m not entirely sure I understand how one copies the data if the original can’t be accessed. Can you please clarify?

**Comment by gsakkis** [comment]
Ok let me try again:

1. On server A I have a nested data structure with np.arrays, such as the following (hypothetical) example:
```
data = {
    "volumes": [
        {
            "x": np.array(...),
            "y": np.array(...),
            "z": np.array(...),
        },
        {
            "x": np.array(...),
            "y": np.array(...),
            "z": np.array(...),
        }
    ],
    "humidity": [
        {
            "a": np.array(...),
            "b": np.array(...),
        },
        {
            "a": np.array(...),
            "b": np.array(...),
        }
    ],
    "date": datetime.date(2020,3,30),
    "verified": True,
    "title": "training data",
}
```
2. I can pickle this as is and unpickle it anywhere, only problem is the size as some of these arrays can be huge. So instead I:
(a) save each `np.array` as zarr
(b) pickle the data structure above after mapping each `np.array` with the respective `zarr.Array`:
```
data = {
    "volumes": [
        {
            "x": zarr.Array(...),
            "y": zarr.Array(...),
            "z": zarr.Array(...),
        },
        {
            "x": zarr.Array(...),
            "y": zarr.Array(...),
            "z": zarr.Array(...),
        }
    ],
    "humidity": [
        {
            "a": zarr.Array(...),
            "b": zarr.Array(...),
        },
        {
            "a": zarr.Array(...),
            "b": zarr.Array(...),
        }
    ],
    "date": datetime.date(2020,3,30),
    "verified": True,
    "title": "training data",
}
```
3. Now I can unpickle the the above, but only as long as the underlying zarr store (directory path, zip file path, etc) is the same as the original one. So if the original arrays are stored under `/path/to/zarr.zip` in server A, if I copy the pickle file to `/tmp` on server B, it blows up during unpickling.
4. With the proposed PR, metadata are not loaded during in `__setstate__` so unpickle doesn't blow up and recreates zarr arrays objects that have a reference to the store of server A. All I need to do then is change the store reference to the valid server B store and then I can access the data just fine.

Hope it's clear now.

**Comment by alimanfoo** [comment]
Hi @gsakkis, 

Maybe this is too much change to your code, but wondering if it would it be possible to take an approach like this...

On server A, create the data, but putting everything in a single zarr hierarchy, e.g.:

```
path = '/original/path/to/zarr.zip'
store = zarr.ZipStore(path, mode='w')
# setup hierarchy
root = zarr.group(store=store)
# create arrays and store data
volumes_x = root.create_dataset('volumes/x', ...)
volumes_x[:] = ...  # store data
volumes_y = root.create_dataset('volumes/y', ...)
volumes_y[:] = ...  # store data
# as above for other arrays volumes/z, humidity/a, etc.
# store metadata
root.attrs.put({
    "date": datetime.date(2020,3,30),
    "verified": True,
    "title": "training data",
}
store.close()
```

Then copy the zip file to a new path on server B, say "/tmp/path/to/zarr.zip". Then send just the new path of the zip file to the worker process on server B. That process can then just run something like:

```
path = '/tmp/path/to/zarr.zip'  # received from server A
store = zarr.ZipStore(path, mode='r')
root = zarr.Group(store=store)
# read data 
volumes_x = root['volumes/x'][:]
volumes_y = root['volumes/y'][:]
...
# read metadata
date = root.attrs['date']
verified = root.attrs['verified']
...
```

I.e., just move the path of the zip file between processes, otherwise encapsulate all data within the zarr hierarchy, no need to pickle or unpickle any data structures?

(Btw I'm pushing back here on zarr code changes because the changes would affect some critical code, but also changes you proposed in #549 would actually change behaviour in some circumstances, e.g., errors would surface later than they do currently if someone attempts to open an array at a location where an array doesn't exist. This seems subtle but a change in behaviour like this would require a major version bump and could have unexpected consequences for current users.)

**Comment by gsakkis** [comment]
Hi @alimanfoo,
thanks for the elaborate example, I'll see if it can be adapted to my context and if it's better or worse than my current approach.

Regardless, I'd like to get a better understanding of where do you see the potential risk of #549. Not sure If you have checked the latest version after my forced pushes but the behaviour now is identical for common Array instantiation; metadata are loaded by default. The only behaviour change is when unpickling a previously pickled instance **and** the store location has changed. How likely is this scenario given that, as far as I can tell from the discussion, moving around pickled zarr arrays is something zarr users rarely do, if ever? 

If you have an alternative, 100% foolproof approach for pickling zarr arrays so that they get one step closer to become a drop-in replacement for numpy arrays, I'd love to hear it :)

**Comment by alimanfoo** [comment]
Thanks @gsakkis, I must have been looking at an earlier commit, I'll take another look asap.

**Comment by gsakkis** [comment]
@jakirkham I'd expect my use case to be simple, even common one, but apparently it's not for some reason:
- I have a bunch of structured data, some of which happen to be (potentially large) numpy arrays.
- These data are produced at place A and are to be consumed at place B.
- That's pretty much it really.

I'm not quite sure what you're proposing; I don't have trouble writing to a ZipStore or any store for that matter; I have trouble reading from it after it is packaged and sent over to the consumer(s). 

@alimanfoo's [suggestion](https://github.com/zarr-developers/zarr-python/issues/548#issuecomment-605984865) seems close towards this end, haven't tried it in my setting yet. The main drawback I see is that it is very Zarr-centric: it requires domain data to fit into (or be adapted to) Zarr's hierarchy and accessed using Zarr's API (e.g. using different syntax for arrays vs metadata). Ideally I'd like to treat zarr as an implementation detail, interchangeable with numpy arrays and (mostly) transparent to the consumer, which is why I'd be in favor of standard python pickle.

**Comment by alimanfoo** [comment]
FWIW I thought of using zarr's hierarchy capability here because it seemed quite close to what you were already building but via some Python data structures, esp. nested dictionaries. But it's not quite the same because you're using both dicts and lists, so appreciate it's not a direct swap-in replacement.

Also FWIW zarr's hierarchy (group) API is intended to match h5py, so it's not entirely zarr-specific. I.e., if you use the group API then it will be familiar to anyone who has used HDF5-like data structures before.

Also FWIW, it sounds like you're putting all the arrays inside a single ZipStore. If you are, then you will be having to decide a path for each array, which will be creating some hierarchical (grouping) structure that I'm guessing mirrors closely the Python nested dicts and lists data structure you're using. So maybe there is an opportunity to just simplify and leverage the grouping structure within the zarr hierarchy, rather than duplicating grouping information into Python objects as well. But maybe it's not that straightforward.

In any case, I'm still planning to look at the updated PR and will give it some more thought.

**Comment by dstansby** [comment]
It seems like pickling an in-memory array works fine:

```python
import pickle, zarr, numpy as np

z = zarr.array(np.arange(10))

with open("walnuts.pickle", "wb") as f:
    pickle.dump(z, f)


with open("walnuts.pickle", "rb") as f:
    pickle.load(f)
```

Is this okay for your use case, instead of going via. a `ZipStore`?
